ipa-core/src/protocol/basics/mul/dzkp_malicious.rs

```diff
@@ -80,20 +80,143 @@ impl<'a, F: Field + DZKPCompatibleField<N>, const N: usize>
 
 #[cfg(all(test, unit_test))]
 mod test {
+    use ipa_step_derive::CompactStep;
+
     use crate::{
         error::Error,
         ff::boolean::Boolean,
+        helpers::{Role, Role::H1},
         protocol::{
-            basics::SecureMul,
-            context::{dzkp_validator::DZKPValidator, Context, DZKPContext, UpgradableContext},
+            basics::{
+                mul::{dzkp_malicious::Field, semi_honest::multiplication_protocol, Replicated},
+                SecureMul,
+            },
+            context::{
+                dzkp_field::DZKPCompatibleField,
+                dzkp_validator::{DZKPValidator, Segment},
+                Context, DZKPContext, DZKPUpgradedMaliciousContext, UpgradableContext,
+            },
             RecordId,
         },
         rand::{thread_rng, Rng},
+        secret_sharing::{replicated::semi_honest::AdditiveShare, SharedValueArray, Vectorizable},
         test_fixture::{Reconstruct, Runner, TestWorld},
     };
 
+    /// This function mirrors `zkp_multiply` except that on party cheats.
+    ///
+    /// The cheating party flips `prss_left`
+    /// which causes a flip in `z_left` computed by the cheating party.
+    /// This manipulated `z_left` is then sent to a different helper
+    /// and included in the DZKP batch.
+    pub async fn multiply_with_cheater<'a, F, const N: usize>(
+        ctx: DZKPUpgradedMaliciousContext<'a>,
+        record_id: RecordId,
+        a: &Replicated<F, N>,
+        b: &Replicated<F, N>,
+        prss: &Replicated<F, N>,
+        cheater: Role,
+    ) -> Result<Replicated<F, N>, Error>
+    where
+        F: Field + DZKPCompatibleField<N>,
+    {
+        let mut prss_left = prss.left_arr().clone();
+        if ctx.role() == cheater {
+            prss_left += <<F as Vectorizable<N>>::Array>::from_fn(|_| F::ONE);
+        };
+
+        let z =
+            multiplication_protocol(&ctx, record_id, a, b, &prss_left, prss.right_arr()).await?;
+        // create segment
+        let segment = Segment::from_entries(
+            F::as_segment_entry(a.left_arr()),
+            F::as_segment_entry(a.right_arr()),
+            F::as_segment_entry(b.left_arr()),
+            F::as_segment_entry(b.right_arr()),
+            F::as_segment_entry(prss.left_arr()),
+            F::as_segment_entry(prss.right_arr()),
+            F::as_segment_entry(z.right_arr()),
+        );
+
+        // add segment to the batch that needs to be verified by the dzkp prover and verifiers
+        ctx.push(record_id, segment);
+
+        Ok(z)
+    }
```

Comment:
This approach leads to a lot of code duplication. Why not just invoke the normal protocol (which already does all of this segment entry stuff), and just mutate the result (apply an additive attack)?

Reply:
I agree. These tests are running against reference implementation and not intended to be formal verification methods that explore all possible states the system can be it. There are better tools than our poor in-memory infra for that.

Modelling the state of malicious adversaries does not seem very useful - there are infinite combinatorial states they can be it. The threat model allows them to delay response infinitely, respond with errors, run Byzantine attacks etc. Instead of focusing on what malicious adversaries **can do**, it may be easier to shift our attention to what honest helpers can see from such entities. This greatly simplifies testing imo, because there are only a few things a malicious actor can do:

1) It may never respond to multiplication requests
2) It can send correct values and tamper the proof
3) It can send corrupted values and proof for correct values
4) It can send corrupted values and proof that matches them

For the first 3 we don't need to duplicate protocols (#1202 covers 2 and 3 and 1 will be covered after #1205). I could be wrong, but test here seems to be checking 3) so we could leverage our new infrastructure capabilities and avoid copy-pasting the segment construction logic.

Both @danielmasny and @andyleiserson expressed interest in covering 4) which is an interesting exercise. I think code duplication is probably wrong for it in the long term as well - we need to figure out how to do it without cloning our MPC circuits.